### PR TITLE
Peter/subscribe package

### PR DESCRIPTION
**Pull request by Ripiter**

### CODE REVIEW

File: Source/Protocol/BTree.cpp [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+
+namespace MQTT {
+	namespace Protocol {
+		BTree::~BTree()
+		{
+			for (int i = 0; i < m_Children.size(); i++)
+				delete m_Children[i];
+
+			for (auto& sub : m_SubClients)
```
Comment (blackout1471) [inline Source/Protocol/BTree.cpp]: Should be "auto*" if you have to use delete

File: Source/Protocol/BTree.h [hunk 2]
```diff
@@ -0,0 +1,77 @@
+#pragma once
+#include <vector>
+#include <string>
+#include "MqttPackages/Packages.h"
+
+namespace MQTT {
+	namespace Protocol {
+		class BTree
+		{
+			std::vector<BTree*> m_Children;
```
Comment (blackout1471) [inline Source/Protocol/BTree.h]: move to the end

File: Source/Protocol/BTree.h [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+			BTree* m_Parent = nullptr;
+			std::vector<unsigned char> m_Topic;
+			std::vector<SubscribeClient*> m_SubClients;
+			std::vector<SubscribeSavedTopic> m_SubSavedTopics;
+			SubscribeTopicWildcardType m_Wildcard = SubscribeTopicWildcardType::NoWildcard;
+
+		public:
```
Comment (blackout1471) [inline Source/Protocol/BTree.h:10]: Comments

File: Source/Protocol/BTree.h [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+			~BTree();
+
+			void AddClient(SubscribeClient* subClient);
+			BTree* GetFullMatch(std::vector<unsigned char> topic, SubscribeTopicWildcardType wildcard);
+			BTree* GetTopicMatch(std::vector<unsigned char> topic);
+
+			std::vector<unsigned char> GetTopic() {
```
Comment (blackout1471) [inline Source/Protocol/BTree.h]: Return as reference, and maybe const

File: Source/Protocol/BTree.h [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+				return m_Topic;
+			}
+
+			SubscribeTopicWildcardType GetWildcard() {
```
Comment (blackout1471) [inline Source/Protocol/BTree.h]: Same as above

File: Source/Protocol/BTree.h [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+				return m_Wildcard;
+			}
+			std::vector<SubscribeClient*>& GetSubClients() {
+				return m_SubClients;
+			}
+
+
+			static BTree* NewBTree(SubscribeClient* subClient, std::vector<unsigned char> topic, SubscribeTopicWildcardType wildcard) {
```
Comment (blackout1471) [inline Source/Protocol/BTree.h]: Move static methods to .cpp

File: Source/Protocol/Converters/ConverterUtility.h [hunk 7]
```diff
@@ -29,6 +29,18 @@ namespace MQTT {
 				{
 					return ControlPackageType((data >> 4) & 0xff);
 				}
+				/*
+				* Converts a int to 2 bytes splitting the value at the 8th bit.
```
Comment (Geefy) [inline Source/Protocol/Converters/ConverterUtility.h]: Dont include this file

File: Source/Protocol/Converters/ConverterUtility.h [hunk 8]
```diff
@@ -29,6 +29,18 @@ namespace MQTT {
...
+				* Returns a vector with 2 values data[0]MSB and data[1]LSB
+				*/
+				static const std::vector<unsigned char> FromIdentifier(int identifier)
```
Comment (blackout1471) [inline Source/Protocol/Converters/ConverterUtility.h]: Wrong naming convention: maybe better to use "FromInt"

File: Source/Protocol/Converters/SubscribeConverter.cpp [hunk 9]
```diff
@@ -0,0 +1,79 @@
+#include "mqttpch.h"
+#include "SubscribeConverter.h"
+#include "ConverterUtility.h"
+namespace MQTT {
+	namespace Protocol {
+		namespace Converters
+		{
+			const SubscribePackage SubscribeConverter::ToPackage(const std::vector<unsigned char>& buffer)
```
Comment (blackout1471) [inline Source/Protocol/Converters/SubscribeConverter.cpp:8]: If possible split into more methods.
Like
* Convertheader()
* ConvertVariableHeader()
* ConvertPayload()

File: Source/Protocol/Converters/SubscribeConverter.h [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+#include "IPackageConverter.h"
+#include "Protocol/MqttPackages/Packages.h"
+namespace MQTT {
+	namespace Protocol {
+		namespace Converters
+		{
+			class SubscribeConverter : public IPackageConverter<SubscribePackage>
+			{
+			public:
+				// Inherited via IPackageConverter
```
Comment (blackout1471) [inline Source/Protocol/Converters/SubscribeConverter.h]: Remove comment

File: Source/Protocol/Managers/SubscribeManager.cpp [hunk 11]
```diff
@@ -0,0 +1,171 @@
+#include "mqttpch.h"
+#include "SubscribeManager.h"
+#include "Protocol/Validators/Rules/Rules.h"
+
+namespace MQTT {
+	namespace Protocol {
+		using namespace Protocol::Validators;
```
Comment (blackout1471) [inline Source/Protocol/Managers/SubscribeManager.cpp]: Don't use 'using namespace' can fuck everything up later in includes

**Review by Geefy — CHANGES_REQUESTED** [review]
Do pull from main

File: Source/Protocol/Managers/SubscribeManager.h [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+#include "Protocol/BTree.h"
+#include "Protocol/Converters/SubscribeConverter.h"
+#include "Protocol/Converters/SubscribeAcknowledgementConverter.h"
+#include "Protocol/Validators/RuleEngine.h"
+#include "Protocol/Validators/Rules/IRule.h"
+
+namespace MQTT {
+	namespace Protocol {
+		class SubscribeManager
+		{
```
Comment (blackout1471) [inline Source/Protocol/Managers/SubscribeManager.h:11]: Move to the end, and use private:

File: Source/Protocol/MqttPackages/SubscribeAcknowledgementPayload.h [hunk 13]
```diff
@@ -0,0 +1,10 @@
+#pragma once
+#include <vector>
+#include "SubscribeAcknowledgementPayloadReturnTypes.h"
+namespace MQTT {
+	namespace Protocol {
+		struct SubscribeAcknowledgementPayload {
+			std::vector<SubscribeAcknowledgementQoS> payload;
```
Comment (blackout1471) [inline Source/Protocol/MqttPackages/SubscribeAcknowledgementPayload.h]: Use capital letter

File: Source/Protocol/MqttPackages/SubscribeClient.h [hunk 14]
```diff
@@ -0,0 +1,16 @@
+#pragma once
+#include <string>
+
+namespace MQTT {
+	namespace Protocol {
+		struct SubscribeClient {
+			std::string ClientID;
```
Comment (blackout1471) [inline Source/Protocol/MqttPackages/SubscribeClient.h]: Move to end of class and private:

**Review by blackout1471 — CHANGES_REQUESTED** [review]

**Review by Geefy — APPROVED** [review]

**Review by blackout1471 — APPROVED** [review]

**Pull request merged**